File: nim/bob/bob.nim
import std/strutils

func isQuestion(s: string): bool =
  var i = s.high
  while i >= 0 and s[i] == ' ':
    dec i
  i >= 0 and s[i] == '?'

func isYelling(s: string): bool =
  result = false
  for c in s:
    if c in {'a'..'z'}:
      return false
    elif c in {'A'..'Z'}:
      result = true

func hey*(s: string): string =
  if isEmptyOrWhitespace(s):
    "Fine. Be that way!"
  elif isQuestion(s):
    if isYelling(s):
      "Calm down, I know what I'm doing!"
    else:
      "Sure."
  elif isYelling(s):
    "Whoa, chill out!"
  else:
    "Whatever."

File: python/bob/_iterations/3/bob.py
def response(hey_bob):
    is_empty = not (hey_bob := hey_bob.strip())
    is_question = not is_empty and hey_bob.endswith("?")
    is_yelling = not is_empty and hey_bob.isupper()

    match (is_empty, is_question, is_yelling):
        case (True, _, _):
            return "Fine. Be that way!"
        case (_, True, True):
            return "Calm down, I know what I'm doing!"
        case (_, True, False):
            return "Sure."
        case (_, False, True):
            return "Whoa, chill out!"
    return "Whatever."
